File: src/main/java/problems/easy/MaximumSubarray/MaximumSubarray.scala
package problems.easy.MaximumSubarray
object SolutionScala {
  def maxSubArray(nums: Array[Int]): Int = {
    // nums.scanLeft(0)((acc, n) => math.max(0, acc + n)).max
    var currentMaximum = -2000000000
    var maximumAtTheEnd = 0
    for (i <- 0 to nums.size - 1) {
      maximumAtTheEnd += nums(i)
      if (currentMaximum < maximumAtTheEnd)
        currentMaximum = maximumAtTheEnd
      if (maximumAtTheEnd < 0)
        maximumAtTheEnd = 0
    }
    currentMaximum
  }
}

File: src/main/java/problems/easy/MaximumSubarray/maximum_subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function (nums) {
  let currentMaximum = Number.MIN_SAFE_INTEGER;
  let maximumAtTheEnd = 0;
  for (let i = 0; i < nums.length; i++) {
    maximumAtTheEnd += nums[i];
    currentMaximum =
      currentMaximum < maximumAtTheEnd ? maximumAtTheEnd : currentMaximum;
    maximumAtTheEnd = maximumAtTheEnd < 0 ? 0 : maximumAtTheEnd;
  }
  return currentMaximum;
};

module.exports = { maxSubArray };
